src/charm.py

```diff
@@ -52,11 +56,17 @@ def __init__(self, *args):
             self._on_nginx_pebble_ready,
         )
 
+        self._s3_storage_data = _S3StorageBackend()
         self.framework.observe(
             self.on.mimir_cluster_relation_changed,  # pyright: ignore
             self._on_mimir_cluster_changed,
         )
 
+        self.framework.observe(
+            self.s3_requirer.on.credentials_changed, self._on_mimir_cluster_changed
```

Comment:
- add a `self._on_s3_requirer_credentials_changed` observer, and from that observer, invoke whatever you need to invoke (and observers shouldn't call one another). see the "Only event handlers should be event-aware" section of https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619